File: lib/client/python/file_exists.py
import os
from google.cloud import storage
from google.api_core import exceptions # 예외 처리를 위해 임포트

os.environ["REQUESTS_CA_BUNDLE"] = "mitmproxy-ca-cert.pem"
os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = "woven-province-411903-b1b12d94b3ac.json"


def list_blobs_in_bucket(bucket_name, output_file="cdn_file_list.txt", timeout=None):
    """지정된 Google Cloud Storage 버킷의 파일 목록을 출력하고 텍스트 파일에 저장합니다."""
    
    # 클라이언트 초기화 - timeout 설정 추가
    storage_client = storage.Client()
    # 클라이언트 타임아웃 설정
    storage_client._http.timeout = timeout if timeout else None

    try:
        # 버킷 내의 모든 blob(파일) 목록 가져오기
        # list_blobs()는 iterator를 반환합니다.
        blobs = storage_client.list_blobs(bucket_name)

        print(f"'{bucket_name}' 버킷의 파일 목록:")
        
        # 파일에 쓰기 위해 파일 열기
        with open(output_file, 'w', encoding='utf-8') as f:
            found_files = False
            for blob in blobs:
                blob_name = blob.name
                print(f"- {blob_name}") # 콘솔에 출력
                f.write(f"{blob_name}\n") # 파일에 저장
                found_files = True

            if not found_files:
                print("  (버킷에 파일이 없습니다)")
                f.write("(버킷에 파일이 없습니다)\n")

    except exceptions.NotFound:
        print(f"오류: 버킷 '{bucket_name}'을(를) 찾을 수 없습니다.")
    except exceptions.Forbidden:
        print(f"오류: 버킷 '{bucket_name}'에 접근할 권한이 없습니다. IAM 권한을 확인하세요.")
    except Exception as e:
        print(f"알 수 없는 오류 발생: {e}")

File: lib/client/go/file_exists.go
package file_exists

import (
	"context"
	"fmt"
	"io"
	"os"
	"time"

	"cloud.google.com/go/storage"
	"google.golang.org/api/iterator"
	"google.golang.org/api/option"
)

func init() {
	// 환경 변수 설정
	os.Setenv("REQUESTS_CA_BUNDLE", "mitmproxy-ca-cert.pem")
	os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", "woven-province-411903-b1b12d94b3ac.json")

	// 기본 버킷 이름과 출력 파일을 지정할 수 있습니다
	bucketName := "your-bucket-name" // 실제 버킷 이름으로 변경하세요
	outputFile := "cdn_file_list.txt"

	// 타임아웃 설정 (선택적)
	var timeout time.Duration // nil 값을 사용하려면 빈 값으로 둡니다

	// 함수 호출
	listBlobsInBucket(bucketName, outputFile, timeout)
}

func listBlobsInBucket(bucketName, outputFile string, timeout time.Duration) {
	ctx := context.Background()

	// 타임아웃 설정이 있는 경우 context에 적용
	if timeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, timeout)
		defer cancel()
	}

	// 스토리지 클라이언트 초기화
	client, err := storage.NewClient(ctx, option.WithCredentialsFile(os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")))
	if err != nil {
		fmt.Printf("스토리지 클라이언트 생성 오류: %v\n", err)
		return
	}
	defer client.Close()

	// 출력 파일 생성
	file, err := os.Create(outputFile)
	if err != nil {
		fmt.Printf("파일 생성 오류: %v\n", err)
		return
	}
	defer file.Close()

	fmt.Printf("'%s' 버킷의 파일 목록:\n", bucketName)

	// 버킷의 객체(blob) 목록 가져오기
	it := client.Bucket(bucketName).Objects(ctx, nil)

	foundFiles := false
	for {
		// 다음 객체 가져오기
		attrs, err := it.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			if isForbiddenError(err) {
				fmt.Printf("오류: 버킷 '%s'에 접근할 권한이 없습니다. IAM 권한을 확인하세요.\n", bucketName)
			} else if isNotFoundError(err) {
				fmt.Printf("오류: 버킷 '%s'을(를) 찾을 수 없습니다.\n", bucketName)
			} else {
				fmt.Printf("객체 목록 가져오기 오류: %v\n", err)
			}
			return
		}

		// 객체 이름 출력 및 파일에 저장
		blobName := attrs.Name
		fmt.Printf("- %s\n", blobName)
		io.WriteString(file, blobName+"\n")
		foundFiles = true
	}

	if !foundFiles {
		fmt.Println("  (버킷에 파일이 없습니다)")
		io.WriteString(file, "(버킷에 파일이 없습니다)\n")
	}
}

// 오류가 권한 거부인지 확인
func isForbiddenError(err error) bool {
	return err.Error() == "storage: bucket access is forbidden"
}

// 오류가 찾을 수 없음인지 확인
func isNotFoundError(err error) bool {
	return err.Error() == "storage: bucket doesn't exist"
}
